File: src/main/kotlin/lec17/Lec17Main.java
package lec17;

import java.util.List;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Lec17Main {

    public static void main(String[] args) {
        List<JavaFruit> fruits = List.of(
                new JavaFruit("사과", 1000),
                new JavaFruit("사과", 2000),
                new JavaFruit("사과", 3000),
                new JavaFruit("바나나", 4000),
                new JavaFruit("바나나", 5000),
                new JavaFruit("포도", 2000),
                new JavaFruit("포도", 3000),
                new JavaFruit("포도", 4000)
        );

        System.out.println(filterFruits(
                        fruits,
                        fruit -> Set.of("사과", "포도").contains(fruit.getName()) && fruit.getPrice() >= 3000
                )
        );
    }

    private static List<JavaFruit> filterFruits(List<JavaFruit> fruits, Predicate<JavaFruit> filter) {
        return fruits.stream()
                .filter(filter)
                .collect(Collectors.toList());
    }

}

File: src/main/kotlin/lec17/Lec17Main.kt
package lec17

fun main() {
    val fruits = listOf(
        Fruit("사과", 1000),
        Fruit("사과", 1000),
        Fruit("사과", 1000),
        Fruit("사과", 1000),
        Fruit("바나나", 1000),
        Fruit("바나나", 1000),
        Fruit("바나나", 1000),
        Fruit("포도", 1000),
        Fruit("포도", 1000),
        Fruit("포도", 1000)
    )

    val isApple = fun(fruit: Fruit): Boolean {
        return fruit.name == "사과"
    }

    val isApple2: (Fruit) -> Boolean = { fruit: Fruit -> fruit.name == "사과" }

    val results = filterFruits(fruits) { it.name == "사과" }
    println(results.toString())
}

private fun filterFruits(fruits: List<Fruit>, filter: (Fruit) -> Boolean): List<Fruit> {
    val result = mutableListOf<Fruit>()
    for (fruit in fruits) {
        if (filter(fruit)) {
            result.add(fruit)
        }
    }
    return result
}